File: src/test/java/cube8540/oauth/authentication/credentials/oauth/token/application/ImplicitTokenGranterTest.java
package cube8540.oauth.authentication.credentials.oauth.token.application;

import cube8540.oauth.authentication.AuthenticationApplication;
import cube8540.oauth.authentication.credentials.AuthorityCode;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedAccessToken;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.oauth2.core.AuthorizationGrantType;

import java.time.Clock;
import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;

import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.ACCESS_TOKEN_ID;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.ACCESS_TOKEN_VALIDITY_SECONDS;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.CLIENT_ID;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.CLIENT_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.RAW_SCOPES;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.TOKEN_CREATED_DATETIME;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.USERNAME;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeClientDetails;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeEmptyAccessTokenRepository;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeTokenIdGenerator;
import static cube8540.oauth.authentication.credentials.oauth.token.application.OAuth2TokenApplicationTestHelper.makeTokenRequest;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

@DisplayName("암묵적 동의를 통한 토큰 부여 테스트")
class ImplicitTokenGranterTest {

    @Test
    @DisplayName("요청 스코프가 null 일때 엑세스 코드 생성")
    void generateAccessTokenWhenScopeOfRequestIsNull() {
        OAuth2ClientDetails clientDetails = makeClientDetails();
        OAuth2AccessTokenRepository repository = makeEmptyAccessTokenRepository();
        OAuth2TokenRequest request = makeTokenRequest();
        ImplicitTokenGranter granter = new ImplicitTokenGranter(makeTokenIdGenerator(ACCESS_TOKEN_ID), repository);

        Clock clock = Clock.fixed(TOKEN_CREATED_DATETIME.toInstant(AuthenticationApplication.DEFAULT_ZONE_OFFSET), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId());
        AbstractOAuth2TokenGranter.setClock(clock);
        when(request.getScopes()).thenReturn(null);

        OAuth2AuthorizedAccessToken accessToken = granter.createAccessToken(clientDetails, request);
        assertEquals(CLIENT_SCOPES, accessToken.getScopes());
        assertAccessToken(accessToken);
    }

    @Test
    @DisplayName("요청 스코프가 비어 있을때 엑세스 코드 생성")
    void generateAccessTokenWhenScopeOfRequestIsEmpty() {
        OAuth2ClientDetails clientDetails = makeClientDetails();
        OAuth2AccessTokenRepository repository = makeEmptyAccessTokenRepository();
        OAuth2TokenRequest request = makeTokenRequest();
        ImplicitTokenGranter granter = new ImplicitTokenGranter(makeTokenIdGenerator(ACCESS_TOKEN_ID), repository);

        Clock clock = Clock.fixed(TOKEN_CREATED_DATETIME.toInstant(AuthenticationApplication.DEFAULT_ZONE_OFFSET), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId());
        AbstractOAuth2TokenGranter.setClock(clock);
        when(request.getScopes()).thenReturn(Collections.emptySet());

        OAuth2AuthorizedAccessToken accessToken = granter.createAccessToken(clientDetails, request);
        assertEquals(CLIENT_SCOPES, accessToken.getScopes());
        assertAccessToken(accessToken);
    }

    @Test
    @DisplayName("요청 스코프가 null이 아닐 때 엑세스 토큰 생성")
    void generateAccessTokenWhenScopeOfRequestIsNotNull() {
        OAuth2ClientDetails clientDetails = makeClientDetails();
        OAuth2AccessTokenRepository repository = makeEmptyAccessTokenRepository();
        OAuth2TokenRequest request = makeTokenRequest();
        ImplicitTokenGranter granter = new ImplicitTokenGranter(makeTokenIdGenerator(ACCESS_TOKEN_ID), repository);

        Clock clock = Clock.fixed(TOKEN_CREATED_DATETIME.toInstant(AuthenticationApplication.DEFAULT_ZONE_OFFSET), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId());
        AbstractOAuth2TokenGranter.setClock(clock);
        when(request.getScopes()).thenReturn(RAW_SCOPES);

        OAuth2AuthorizedAccessToken accessToken = granter.createAccessToken(clientDetails, request);
        Set<AuthorityCode> exceptedScopes = RAW_SCOPES.stream().map(AuthorityCode::new).collect(Collectors.toSet());
        assertEquals(exceptedScopes, accessToken.getScopes());
        assertAccessToken(accessToken);
    }

    private void assertAccessToken(OAuth2AuthorizedAccessToken accessToken) {
        assertEquals(ACCESS_TOKEN_ID, accessToken.getTokenId());
        assertEquals(USERNAME, accessToken.getUsername());
        assertEquals(CLIENT_ID, accessToken.getClient());
        assertEquals(AuthorizationGrantType.IMPLICIT, accessToken.getTokenGrantType());
        assertEquals(TOKEN_CREATED_DATETIME, accessToken.getIssuedAt());
        assertEquals(TOKEN_CREATED_DATETIME.plusSeconds(ACCESS_TOKEN_VALIDITY_SECONDS), accessToken.getExpiration());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/token/application/ImplicitTokenGranterTest.kt
package cube8540.oauth.authentication.oauth.token.application

import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.oauth.error.InvalidRequestException
import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.oauth.security.OAuth2RequestValidator
import cube8540.oauth.authentication.oauth.security.OAuth2TokenRequest
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenRepository
import cube8540.oauth.authentication.oauth.token.domain.OAuth2TokenId
import cube8540.oauth.authentication.oauth.token.domain.OAuth2TokenIdGenerator
import cube8540.oauth.authentication.oauth.token.domain.PrincipalUsername
import cube8540.oauth.authentication.security.AuthorityCode
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Test
import org.springframework.security.oauth2.core.OAuth2ErrorCodes

class ImplicitTokenGranterTest {

    private val tokenIdGenerator: OAuth2TokenIdGenerator = mockk {
        every { generateTokenValue() } returns OAuth2TokenId("tokenId")
    }
    private val repository: OAuth2AccessTokenRepository = mockk(relaxed = true) {
        every { save(any()) } returnsArgument 0
    }
    private val validator: OAuth2RequestValidator = mockk()

    private val granter = ImplicitTokenGranter(tokenIdGenerator, repository)

    init {
        granter.tokenRequestValidator = validator
    }

    @Test
    fun `create access token when request username is null`() {
        val clientDetails: OAuth2ClientDetails = mockk(relaxed = true)
        val tokenRequest: OAuth2TokenRequest = mockk {
            every { username } returns null
        }

        val thrown = catchThrowable { granter.createAccessToken(clientDetails, tokenRequest) }
        assertThat(thrown).isInstanceOf(InvalidRequestException::class.java)
        assertThat((thrown as InvalidRequestException).error.errorCode).isEqualTo(OAuth2ErrorCodes.INVALID_REQUEST)
    }

    @Test
    fun `create access token when request scope is null`() {
        val clientDetails: OAuth2ClientDetails = mockk(relaxed = true) {
            every { clientId } returns "clientId"
            every { scopes } returns setOf("client-scope-1", "client-scope-2", "client-scope-3")
        }
        val tokenRequest: OAuth2TokenRequest = mockk {
            every { username } returns "username"
            every { scopes } returns null
        }

        every { validator.validateScopes(clientDetails, null) } returns true

        val result = granter.createAccessToken(clientDetails, tokenRequest)
        assertThat(result.username).isEqualTo(PrincipalUsername("username"))
        assertThat(result.client).isEqualTo(OAuth2ClientId("clientId"))
        assertThat(result.scopes)
            .isEqualTo(setOf(AuthorityCode("client-scope-1"), AuthorityCode("client-scope-2"), AuthorityCode("client-scope-3")))
    }

    @Test
    fun `create access token when request scope is empty`() {
        val clientDetails: OAuth2ClientDetails = mockk(relaxed = true) {
            every { clientId } returns "clientId"
            every { scopes } returns setOf("client-scope-1", "client-scope-2", "client-scope-3")
        }
        val tokenRequest: OAuth2TokenRequest = mockk {
            every { username } returns "username"
            every { scopes } returns emptySet()
        }

        every { validator.validateScopes(clientDetails, emptySet()) } returns true

        val result = granter.createAccessToken(clientDetails, tokenRequest)
        assertThat(result.username).isEqualTo(PrincipalUsername("username"))
        assertThat(result.client).isEqualTo(OAuth2ClientId("clientId"))
        assertThat(result.scopes)
            .isEqualTo(setOf(AuthorityCode("client-scope-1"), AuthorityCode("client-scope-2"), AuthorityCode("client-scope-3")))
    }

    @Test
    fun `create access token when request scope is allowed`() {
        val clientDetails: OAuth2ClientDetails = mockk(relaxed = true) {
            every { clientId } returns "clientId"
            every { scopes } returns setOf("client-scope-1", "client-scope-2", "client-scope-3")
        }
        val tokenRequest: OAuth2TokenRequest = mockk {
            every { username } returns "username"
            every { scopes } returns setOf("request-scope-1", "request-scope-2", "request-scope-3")
        }

        every { validator.validateScopes(clientDetails, setOf("request-scope-1", "request-scope-2", "request-scope-3")) } returns true

        val result = granter.createAccessToken(clientDetails, tokenRequest)
        assertThat(result.username).isEqualTo(PrincipalUsername("username"))
        assertThat(result.client).isEqualTo(OAuth2ClientId("clientId"))
        assertThat(result.scopes).isEqualTo(setOf(AuthorityCode("request-scope-1"), AuthorityCode("request-scope-2"), AuthorityCode("request-scope-3")))
    }
}